File: trunk/S2SH/WebRoot/dojo/code/remote-scripting/remote-scripting.js
/***
 * Excerpted from "Mastering Dojo",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/rgdojo for more book information.
***/

function example1(){
  dojo.xhrGet({
    url: "demo/id1",
    load: function(response){alert(response);},
    error: function(error){alert(error.message);}
  });
}



function example2(){
  dojo.xhrGet({
    url: "demo/id1",
    handle: function(response){
      if (response instanceof Error) {
        //failed...
        alert("failed: " + response.message);
      }
      else {
        //success...
        alert('succeeded: "' + response + '"');
      }
   }
  });
}



function handler1(response, ioArgs){
  var error= response instanceof Error;
  var responseText= error ? response.message : response;
  var resultNode= dojo.byId("result");
  resultNode.innerHTML= responseText.replace(/</g, "&lt;");
  dojo.toggleClass(resultNode, "error", error);
  dojo.byId("objects").innerHTML= 
    dumpObject({"this": this, ioArgs: ioArgs});
}

function example3(){
  dojo.xhrGet({
    url: "demo/id1",
    handle: handler1
  });
}



function example4(){
  //get some variables that we'll use in the handler function...
  var targetNode= dojo.byId("result");
  
  //make a handler closed on the variables we made...
  function handler2(response){
    var error= response instanceof Error;
    var responseText= error ? response.message : response;
    targetNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(targetNode, "error", error);
    
    //note: NOT dumping anything...
    dojo.byId("objects").innerHTML= "";
  }
  
  //make the XHR call...
  dojo.xhrGet({
    url: "demo/id1",
    handle: handler2
  });
}



function example5(){
  function handler3(response){
    var error= response instanceof Error;
    var responseText= error ? response.message : response;
    this._user.targetNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(this._user.targetNode, "error", error);
    
    //note: ONLY dumping this...
    dojo.byId("objects").innerHTML= 
      dumpObject({"this": this});
  }
  
  //make the XHR call...
  dojo.xhrGet({
    url: "demo/id1",
    handle: handler3,
    _user: {targetNode: dojo.byId("result")}
  });
}




function handler4(response, ioArgs){
  var error= response instanceof Error;
  var responseText= error ? response.message : ioArgs.xhr.responseText;
  var resultNode= dojo.byId("result");
  resultNode.innerHTML= responseText.replace(/</g, "&lt;");
  dojo.toggleClass(resultNode, "error", error);
  dojo.byId("objects").innerHTML= 
    dumpObject({response: response});
}

function example6(){
  dojo.xhrGet({
    url: "demo/id2",
    handleAs: "json",
    handle: handler4
  });
}



function example7(){
  dojo.xhrGet({
    url: "demo/id3",
    handleAs: "json-comment-filtered",
    handle: handler4
  });
}



function example8(){
  dojo.xhrGet({
    url: "demo/id2",
    handleAs: "json-comment-optional",
    handle: handler4
  });
}

function example9(){
  dojo.xhrGet({
    url: "demo/id3",
    handleAs: "json-comment-optional",
    handle: handler4
  });
}


function example10(){
  dojo.xhrGet({
    url: "demo/id4",
    handleAs: "javascript",
    handle: handler4
  });
  //returns "Number(dojo.byId("exId").value) * 10"
}



function example11(){
  function handler5(response, ioArgs){
    var error= response instanceof Error;
    var responseText= error ? response.message : ioArgs.xhr.responseText;
    var resultNode= dojo.byId("result");
    resultNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(resultNode, "error", error);
    dojo.byId("objects").innerHTML= 
      dumpObject({response: response});
  }

  dojo.xhrGet({
    url: "demo/id5",
    handleAs: "xml",
    handle: handler5
  });
}


function example12(){
  function handler12(response, ioArgs){
    var error= response instanceof Error;
    var responseText= error ? response.message : ioArgs.xhr.responseText;
    var resultNode= dojo.byId("result");
    resultNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(resultNode, "error", error);
    dojo.byId("objects").innerHTML= 
      dumpObject({"this": this, ioArgs: ioArgs}, {maxCount: 15});
  }
  

  dojo.xhrGet({
    url: "demo/id6",
    user: "john",
    password: "open-sesame",
    content: {
      param1: "someParamValue",
      param2: 2.7183
    },
    contentType: "text/plain",
    headers: {
      myKey: "someValue",
      anotherKey: 3.1416
    },
    synch: true,
    preventCache: true,
    handleAs: "json-comment-filtered",
    handle: handler12
  });
}



function example13(){
  dojo.xhrGet({
    form: "exForm",
    handleAs: "json-comment-filtered",
    handle: handler4
  });
}



function example14(){
  dojo.xhrGet({
    form: "exForm",
    content: {sex: "m", anotherParam: 3.1416},
    handleAs: "json-comment-filtered",
    handle: handler4
  });
}


function example99(){
  dojo.xhrGet({
    url: "demo/ex2",
    timeout: 50,
    handle: handler1
  });
}


function example15(){
  function loadScript(url, checkString){

    var element = dojo.doc.createElement("script");
    element.type = "text/javascript";
    element.src = url;
    dojo.doc.getElementsByTagName("head")[0].appendChild(element);

    var intervalId= setInterval(check, 50);
    
    function check(){
      if (eval("typeof(" + checkString + ") != 'undefined'")) {
        clearInterval(intervalId);
        o = dojo.getObject(checkString);
        dojo.byId("result").innerHTML = o.result.replace(/</g, "&lt;");
      }
    }
  }
  
  loadScript("demo/id8", "id8Object")
}


function snip102(){

id8Object= {};
id8Object.result= "exercise15--hello, world";

}


function example16(){
  function loadScript(url, callback){
    dojo.setObject(callback, function(o){
      dojo.byId("result").innerHTML = o.result.replace(/</g, "&lt;");
    })
        
    var element = dojo.doc.createElement("script");
    element.type = "text/javascript";
    element.src = url + "?callbackName=" + callback;
    dojo.doc.getElementsByTagName("head")[0].appendChild(element);
  }
  
  loadScript("demo/id9", "exercise16.myCallback")
}


function snip104(){

exercise16.myCallback({"result": "hello, world"});

}


dojo.require("dojo.io.script");
function example17(){
  
  function handler17(response, ioArgs){
    var error= response instanceof Error;
    var responseText= error ? response.message : id8Object.result;
    var resultNode= dojo.byId("result");
    resultNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(resultNode, "error", error);
    dojo.byId("objects").innerHTML= 
      dumpObject({response: response, ioArgs: ioArgs});
  }
  
  dojo.io.script.get({
    url:"demo/id8",
    checkString:"id8Object",
    handle:handler17
  });
}




function example18(){
  dojo.require("dojo.io.script");
  
  function handler18(response, ioArgs){
    var error= response instanceof Error;
    var responseText= error ? response.message : response.result;
    var resultNode= dojo.byId("result");
    resultNode.innerHTML= responseText.replace(/</g, "&lt;");
    dojo.toggleClass(resultNode, "error", error);
    dojo.byId("objects").innerHTML= 
      dumpObject({"response": response, ioArgs: ioArgs});
  }
  
  dojo.io.script.get({
    url:"demo/id9",
    callbackParamName:"callbackName",
    handle:handler18
  });
}



dojo.require("dojo.io.iframe");
function handler20(response, ioArgs){
  if (!(response instanceof Error)) {
    dojo.byId("objects").innerHTML= 
      dumpObject({response: response, ioArgs: ioArgs});
  } else {
    var resultNode= dojo.byId("result");
    resultNode.innerHTML= response.message;
    dojo.toggleClass(resultNode, "error", true);
    dojo.byId("objects").innerHTML= 
      dumpObject({response: response, ioArgs: ioArgs});    
  }
}

function example20(){
  dojo.io.iframe.send({
    form: "exForm",
    url: "demo/id11",
    method: "post",
    handleAs: "json",
    handle: handler20
  });
}


/*

<html>
  <body>
      <textarea>{result: "OK"}</textarea>
  </body>
</html>

*/

/*

<script 
  type="text/javascript" 
  src="http://o.aolcdn.com/dojo/1.1.0/dojo/dojo.xd.js" 
  djConfig=
    "dojoBlankHtmlUrl: 'http://www.someCompany.com/dojo/resources/blank.html'">
</script>

*/



function example21(){
  
  function add(x, y) {
    var result;
    dojo.xhrPut({
      url: "http://calculator.com",
      content: {proc: "add", op1: x, op2: y},
      synch: true,
      handle: function(result_) {
        result= result_;
      }
    });
    return result;    
  }
  
  var one_plus_two= add(1, 2);
  
  alert("add(1, 2)= " + one_plus_two);
  
}


dojo.require("dojo.rpc.JsonService");
dojo.require("dojo.rpc.JsonpService");
function example22(){
  
  var arithmeticService = {
    serviceURL: "demo/id7",
    methods: [{
      name: "add",
      parameters: [{
        name: "op1",
        type: "number"
      }, {
        name: "op2",
        type: "number"
      }]
    }, {
      name: "subtract",
      parameters: [{
        name: "op1",
        type: "number"
      }, {
        name: "op2",
        type: "number"
      }]
    }]
  };
  
  
  
  //don't forget the keyword new!
  var calculator = new dojo.rpc.JsonService(arithmeticService);
  
  
  
  calculator.add(3, 4).addCallback(function(result){
    alert("3 + 4= " + result);
  });
  
  return;
  
  var junk= 
  
  {"params":[3,4],"method":"add","id":1}
  
  ;

  if (false) {
    
    //resultObject is the JSON object returned...
    if (resultObject.error!=null) {
      call_controlling_deferred.errback(parseError(resultObject));
    } else {
      call_controlling_deferred.callback(owning_rpc_dispatcher.parseResults(obj));
    }
    
  }

  if (false) {
    
    //resultObject is the JSON object returned...
    function parseResults(resultObject) {
      if (dojo.isObject(resultObject)) {
        if ("result" in resultObject) {
          return resultObject.result;
        }
        if ("Result" in resultObject) {
          return resultObject.Result;
        }
        if ("ResultSet" in resultObject) {
          return resultObject.ResultSet;
        }
      }
      return resultObject;
    }
    
  }

  if (false) {
    
    //resultObject is the JSON object returned...
    function parseError(resultObject) {
      var error;
      if (typeof resultObject.error == 'object') {
        error = new Error(resultObject.error.message);
        error.code = resultObject.error.code;
        error.error = resultObject.error.error;
      } else {
        error = new Error(resultObject.error);
      }
      error.id = resultObject.id;
      error.errorObject = resultObject;
      return error;
    }
    
  }
  
  
  var calculator = new dojo.rpc.JsonService(arithmeticService);
  calculator.parseResults= function(resultObject){
    return resultObject.arithmeticServiceResult;
  }
  
  
  
  var calculator = new dojo.rpc.JsonpService(arithmeticService);
  calculator.callbackParamName = "arithmeticService-callback-name";
  
  
  calculator.add({op1: 3, op2: 4}).addCallback(function(result){
    alert("3 + 4= " + result);
  });
  
}

File: trunk/S2SH/WebRoot/dojo/code/remote-scripting/remote-scripting.rb
#---
# Excerpted from "Mastering Dojo",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/rgdojo for more book information.
#---
require 'webrick'
require 'uri'

include WEBrick

class MasteringDojoRemoteScripting < HTTPServlet::AbstractServlet
  def do_GET(request, response)
    id= request.request_uri.path.match('[^/]+$')[0]
    self.send(id, request, response, "myTest")
    response.status= 200
    return
  rescue
    response.status= 400
  end

  def do_POST(request, response)
    id= request.request_uri.path.match('[^/]+$')[0]
    self.send(id, request, response, "myTest")
    response.status= 200
    return
  rescue
    response.status= 400
  end
  
  def id1(request, response, test)
    response.body= "hello, world"
    response['Content-Type']= 'text/plain'
  end  
    
  def id2(request, response, test)
    response.body= '{"firstName": "George", "lastName": "Bush", "address": {"street": "1600 Pennsylvania Avenue NW", "city": "Washington", "state": "DC", "zip": "20500"}}';
    response['Content-Type']= 'text/plain'
  end    
    
  def id3(request, response, test)
    response.body= '/*{"firstName": "George", "lastName": "Bush", "address": {"street": "1600 Pennsylvania Avenue NW", "city": "Washington", "state": "DC", "zip": "20500"}}*/';
    response['Content-Type']= 'text/plain'
  end 
      
  def id4(request, response, test)
    response.body= 'Number(dojo.byId("exId").value) * 10'
    response['Content-Type']= 'text/plain'
  end 
      
  def id5(request, response, test)
    response.body= '<contact><firstName>George</firstName><lastName>Bush</lastName><address street="1600 Pennsylvania Avenue NW" city="Washington" state="DC" zip="20500" /></contact>'
    response['Content-Type']= 'text/xml'
  end   
      
  def id6(request, response, test)
    response.body= '/*{"firstName": "George", "lastName": "Bush", "address": {"street": "1600 Pennsylvania Avenue NW", "city": "Washington", "state": "DC", "zip": "20500"}}*/';
    response['Content-Type']= 'text/plain'
  end 
      
  def id7(request, response, test)
    params= []
    request.query.each{|k, v|
      params.push('"' + k.to_s + '": "' + v.to_s + '"')
    }
    response.body= "{result:{" + params.join(",") + "}}"
    # response.body= '{"result": "' + request.body + '"}'
    response.body= '{"result": "7"}'
    response['Content-Type']= 'text/plain'
  end
  
  def id8(request, response, test)
    response.body= %q{
      id8Object= {};
      id8Object.result= "exercise15--hello, world";
    }
    response['Content-Type']= 'text/javascript'
  end 
  
  def id9(request, response, test)
    response.body= %Q{
      #{request.query["callbackName"]}({"result": "exercise16--hello, world"});
    }
    response['Content-Type']= 'text/javascript'
  end
  
  def id10(request, response, test)
    sleep(5)
    STDOUT << request.body
    response.body= %Q{<div id="result"><p>hello, world from iframe</p></div>'}
    
    response['Content-Type']= 'text/html'
  end
  
  def id11(request, response, test)
    STDOUT << request.body
    response.body= %Q{
      <html>
        <body>
            <textarea>{result: "OK"}</textarea>
        </body>
      </html>
    }
    response['Content-Type']= 'text/html'
  end
  
  def id12(request, response, test)
    if (request.query["escape"])
      response.body= %Q{
        <html>
          <head>
            <script type="text/javascript">
              function rawldLoad(){
                value = window["document"].getElementsByTagName("textarea")[0].value; //text
                this.parent.rawld= value;
                debugger;
                alert(value);
              }
            </script>
          </head>
          <body onLoad="rawldLoad();">
              <textarea>{"left-angle-bracket": "&amp;lt;", "amperstand": "&amp;amp;", "close": "&lt;/textarea>"}</textarea>
          </body>
        </html>
      }
    else
      response.body= %Q{
        <html>
          <head>
          </head>
          <body>
              <textarea>{"left-angle-bracket": "<", "amperstand": "&", "close": "</textarea>"}</textarea>
          </body>
        </html>
      }
    end
    response['Content-Type']= 'text/html'
  end
  
  def ex50(request, response, test)
    response.body= %Q{<p>response: #{request.query["clickId"]}</p><p>hello, world</p>}
    response['Content-Type']= 'text/plain'
  end
  
  def ex3(request, response, test)
    sleep(5)
    response.body= "pause 3s, then...hello, world"
    response['Content-Type']= 'text/plain'
  end
end

server= HTTPServer.new(:Port=>8002, :DocumentRoot=>'c:/dev/dojo-book/code')
server.mount('/remote-scripting/demo', MasteringDojoRemoteScripting)
server.mount("/dojoroot", HTTPServlet::FileHandler, "C:/dev/dojo/dojo-release-1.1.0-src")

['TERM', 'INT'].each do |signal|
  trap(signal){server.shutdown}
end

server.start
